### PR TITLE
Break up core.py

### PR DESCRIPTION
**Pull request by marcromeyn**
### Goals :soccer:
Currently `tf.core` is very large and therefore hard to navigate. It also leads to potential issues for our docs. This PR addresses this.

### Implementation Details :construction:
Implementation is based on [this](https://adamj.eu/tech/2021/05/13/python-type-hints-how-to-fix-circular-imports/) article.

### Testing Details :mag:
No extra tests, since it’s just moving classes around.

### CODE REVIEW

File: merlin/models/tf/blocks/core/base.py [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+
+
+def _output_metrics(metrics):
+    if len(metrics) == 1:
+        return metrics[list(metrics.keys())[0]]
+
+    return metrics
+
+
+def right_shift_layer(self, other):
```
Comment (benfred) [inline merlin/models/tf/blocks/core/base.py:616]: Do we still need this functionality?

**Review by benfred — COMMENTED** [review]
this is looking good!

I do think we can probably break up tf/blocks/core/combinators.py more though - its still clocking in at over 1K lines of code, and contains many different classes that could be split off into their own files. What do you think?

Left some other minor comments/questions below

**Review by gabrielspmoreira — COMMENTED** [review]
This restructuring PR looks good to me. I have just a few general questions about the PR.

1) I see mypy is currently not enabled in pre-commit? What is our plan to enable it? How could we ensure that we are adding proper hints (e.g. # type: ignore) where needed?

2) What should it be the criteria for us to decide which classes we should expose to the highest level TF API (https://github.com/NVIDIA-Merlin/models/blob/tf/refactor/merlin/models/tf/__init__.py)

File: merlin/models/tf/blocks/core/aggregation.py [hunk 2]
```diff
@@ -272,28 +272,27 @@ def get_config(self):
...
+    else:
+        if not len(inputs) == 2:
+            raise ValueError(f"Expected 2 inputs, got {len(inputs)}")
+        left, right = tuple(inputs.values())
+    outputs = self.super().__call__(left, right, **kwargs)  # type: ignore
+
+    return outputs
+
+
+TupleAggregation.__call__ = _tuple_aggregation_call  # type: ignore
```
[the remark this answers is not in the document]
Comment (marcromeyn) [reply merlin/models/tf/blocks/core/aggregation.py]: Added a comment here. Somehow if you overwrite __call__ in a tensorflow Layer, mypy crashes...

**Review by gabrielspmoreira — APPROVED** [review]

**Pull request merged**